moveit_plugins/moveit_simple_controller_manager/src/moveit_simple_controller_manager.cpp

```diff
@@ -35,131 +35,166 @@
 #include <moveit_simple_controller_manager/gripper_controller_handle.h>
 #include <moveit_simple_controller_manager/follow_joint_trajectory_controller_handle.h>
-#include <moveit/utils/xmlrpc_casts.h>
+// #include <moveit/utils/xmlrpc_casts.h>
+#include <boost/algorithm/string/join.hpp>
 #include <pluginlib/class_list_macros.hpp>
 #include <algorithm>
 #include <map>
 
-using namespace moveit::core;
+static const rclcpp::Logger LOGGER = rclcpp::get_logger("moveit_simple_controller_manager");
 
-const std::string LOGNAME("SimpleControllerManager");
+// TOOD(JafarAbdi): Should be in moveit/utils .?
+bool isSubParameter(const std::string& ns, const std::map<std::string, rclcpp::Parameter>& parameters,
+                    const std::vector<std::string>& keys)
+{
+  std::vector<std::string> missing;
+  for (const std::string& key : keys)
+  {
+    if (parameters.find(key) == parameters.end())
+      missing.push_back(key);
+  }
+  if (!missing.empty())
+  {
+    RCLCPP_WARN_STREAM(LOGGER, ns << " is not a namespace with keys " << boost::join(keys, ",") << " (misses "
+                                  << boost::join(missing, ",") << ")");
+    return false;
+  }
+  return missing.empty();
+}
 
 namespace moveit_simple_controller_manager
 {
 class MoveItSimpleControllerManager : public moveit_controller_manager::MoveItControllerManager
 {
 public:
-  MoveItSimpleControllerManager() : node_handle_("~")
+  MoveItSimpleControllerManager() = default;
+
+  ~MoveItSimpleControllerManager() override = default;
+
+  void initialize(const rclcpp::Node::SharedPtr& node) override
   {
-    if (!node_handle_.hasParam("controller_list"))
+    node_ = node;
+    if (!node_->has_parameter("controller_names"))
     {
-      ROS_ERROR_STREAM_NAMED(LOGNAME, "No controller_list specified.");
+      RCLCPP_ERROR_STREAM(LOGGER, "No controller_names specified.");
       return;
     }
-
-    XmlRpc::XmlRpcValue controller_list;
-    node_handle_.getParam("controller_list", controller_list);
-    if (!isArray(controller_list))
+    rclcpp::Parameter controller_names_param;
+    node_->get_parameter("controller_names", controller_names_param);
+    if (controller_names_param.get_type() != rclcpp::ParameterType::PARAMETER_STRING_ARRAY)
     {
-      ROS_ERROR_NAMED(LOGNAME, "Parameter controller_list should be specified as an array");
+      RCLCPP_ERROR(LOGGER, "Parameter controller_names should be specified as a string array");
       return;
     }
-
+    std::vector<std::string> controller_names = controller_names_param.as_string_array();
     /* actually create each controller */
-    for (int i = 0; i < controller_list.size(); ++i)  // NOLINT(modernize-loop-convert)
+    for (const std::string& controller_name : controller_names)
     {
-      if (!isStruct(controller_list[i], { "name", "joints", "action_ns", "type" }))
+      std::map<std::string, rclcpp::Parameter> controller_params;
+      node_->get_parameters(controller_name, controller_params);
```

Comment:
Or `node_->get_node_parameters_interface()->get_parameters_by_prefix(controller_name, controller_params);` .?

`node_->get_parameters` is supposed to be called with map of strings to a type (bool, int, string, ..etc) but it works with having `rclcpp::Parameter` not sure if it's hacky

Reply:
I think it's supposed to work like that